### DISCUSSION TITLE
Take into account wildignore setting

### DISCUSSION THREAD

**Discussion by cassepipe** [nvim-tree/nvim-tree.lua#2169] [Ideas]
I went and tried to see if nvimtree was taking wildignore options into account and it does not.

It would make a lot of sense if it did.

What do you think ?

For reference :
```
	*'wildignore'* *'wig'*
'wildignore' 'wig'	string	(default "")
			global
	A list of file patterns.  A file that matches with one of these
	patterns is ignored when expanding |wildcards|, completing file or
	directory names, and influences the result of |expand()|, |glob()| and
	|globpath()| unless a flag is passed to disable this.
	The pattern is used like with |:autocmd|, see |autocmd-pattern|.
	Also see 'suffixes'.
	Example:  
		:set wildignore=*.o,*.obj
 	The use of |:set+=| and |:set-=| is preferred when adding or removing
	a pattern from the list.  This avoids problems when a future version
	uses another default.
```

**Comment by alex-courtis** [comment]
Interesting.

What would wildignore apply to?

**Reply by cassepipe** [reply]
The way I thought about it was for the file explorer to hide those files by default.
If you you want those files not to appear in vim's completions, you probably don't want them in the file explorer either right ?

I actually though about this  because I saw in the shortcuts that there was both `H` to toggle dotfiles and `U` to toggle `Hidden` (?) in the `g?` shortcut listing so I thought oh surely `Toggle Hidden` is to show files ignored by vim's `wildignore` option but nope, it seems to be something else.

**Reply by alex-courtis** [reply]
That could work. wildignore could be evaluate by `expand` e.g. `vim.fn.expand("/path/to/README.md", false, true)`

Hidden applies to dotfiles.

Please raise a PR and include some real world test cases.

**Reply by cassepipe** [reply]
If `Toggle Hidden` applies to dotfiles... What about `Toggle Dotfiles` ?

I am looking at the codebase and I will see if it is simple enough for me to make a PR :+1: 

I have started learning a bit of lua and I am currently reading : https://neovim.io/doc/user/lua-guide.html#lua-guide-options

**Reply by gegoune** [reply]
Toggling dotfiles already exists (H by default).

**Reply by cassepipe** [reply]
@gegoune Yes `H` is  for`Toggle Dotfiles` I mention just above and `U` is for `Toggle Hidden`. You can see both with `g?` when focusing the explorer. If you happen to know  the diefference, I am interested.

What do you think of this idea of taking into account whether that option is set ?

**Reply by alex-courtis** [reply]
We should keep it simple and keep all the filters orthogonal.

We could add new API `api.tree.toggle_wildignore_filter` and the functionality to `filters.lua`.

**Reply by alex-courtis** [reply]
Alternatively we could to just add `nvim-tree.filters.custom_function` and give full power to the user instead.

`custom_function` would take an (absolute) path - that's all the information we have.

**Reply by cassepipe** [reply]
Ok, so from what I understand from the `docs` page `Toggle Hidden` toggles displaying files based on a custom filter.

I still think it would be a pity a to require the user to write a custom filter function to take `wildignore` into account since the very existence of that option is to have `vim` ignore files, shouldn't be the default ?

**Reply by alex-courtis** [reply]
> I still think it would be a pity a to require the user to write a custom filter function to take `wildignore` into account since the very existence of that option is to have `vim` ignore files, shouldn't be the default ?

If we were building the tree from scratch today, yes. Unfortunately that would break existing users, so we may only add opt-in functionality.
